From 241da6b5588d965d9e5dfef153577067b11bf6b1 Mon Sep 17 00:00:00 2001
From: PratikBhusal <PratikBhusal@users.noreply.github.com>
Date: Wed, 18 Dec 2019 00:55:29 -0600
Subject: [PATCH] Init image mode configs from ~/.Xresources file

The following setitngs from config.h can now be loaded from your
Xresources config file:

GAMMA_MAX    ( default: sxiv.gammaMax            )
GAMMA_RANGE  ( default: sxiv.gammaRange          )
PAN_FRACTION ( default: sxiv.panFraction         )
ANTI_ALIAS   ( default: sxiv.enableAntiAlias     )
ALPHA_LAYER  ( default: sxiv.startWithAlphaLayer )

Note: you do need to get a new copy of config.h as this patch modifies
config.def.h
---
 config.def.h | 26 +++++++++++++++++++++-----
 image.c      |  4 ++++
 2 files changed, 25 insertions(+), 5 deletions(-)

diff --git a/config.def.h b/config.def.h
index f8f1df5..c2fa959 100644
--- a/config.def.h
+++ b/config.def.h
@@ -39,21 +39,37 @@ enum { SLIDESHOW_DELAY = 5 };
 /* gamma correction: the user-visible ranges [-GAMMA_RANGE, 0] and
  * (0, GAMMA_RANGE] are mapped to the ranges [0, 1], and (1, GAMMA_MAX].
  * */
-static const double GAMMA_MAX   = 10.0;
-static const int    GAMMA_RANGE = 32;
+static double GAMMA_MAX   = 10.0;
+static int    GAMMA_RANGE = 32;
 
 /* command i_scroll pans image 1/PAN_FRACTION of screen width/height */
-static const int PAN_FRACTION = 5;
+static int PAN_FRACTION = 5;
 
 /* if false, pixelate images at zoom level != 100%,
  * toggled with 'a' key binding
  */
-static const bool ANTI_ALIAS = true;
+static bool ANTI_ALIAS = true;
 
 /* if true, use a checkerboard background for alpha layer,
  * toggled with 'A' key binding
  */
-static const bool ALPHA_LAYER = false;
+static bool ALPHA_LAYER = false;
+
+/*
+ * Xresources preference(s) for image mode settings to load at startup. If you
+ * do not want to use Xresources for the image mode configuration, uncomment the
+ * line below:
+ */
+/* #undef _XRESOURCES_IMAGE */
+#if defined(_XRESOURCES_CONFIG) && defined(_XRESOURCES_IMAGE)
+    static const ResourcePref image_resources[] = {
+        { "gammaMax",            DOUBLE, &GAMMA_MAX    },
+        { "gammaRange",          INT,    &GAMMA_RANGE  },
+        { "panFraction",         INT,    &PAN_FRACTION },
+        { "enableAntiAlias",     BOOL,   &ANTI_ALIAS   },
+        { "startWithAlphaLayer", BOOL,   &ALPHA_LAYER  },
+    };
+#endif /* _XRESOURCES_CONFIG */
 
 #endif
 #ifdef _THUMBS_CONFIG
diff --git a/image.c b/image.c
index 1230750..c043236 100644
--- a/image.c
+++ b/image.c
@@ -46,6 +46,10 @@ static int zoomdiff(img_t *img, float z)
 
 void img_init(img_t *img, win_t *win)
 {
+	#if defined(_XRESOURCES_CONFIG) && defined(_XRESOURCES_IMAGE)
+		get_xresources(image_resources, ARRLEN(image_resources));
+	#endif /* _XRESOURCES_CONFIG */
+
 	zoom_min = zoom_levels[0] / 100.0;
 	zoom_max = zoom_levels[ARRLEN(zoom_levels) - 1] / 100.0;
 
-- 
2.24.1

